File: flutter/test/failures/sellwild_failures_test.dart
// The logFailure shell (FAILURES.md 3.4): context, error mapping, the
// never-throw wrapper, the reentrancy guard, the debug echo, kill switches
// and the default route through SellwildAPIClient.instance.

import 'dart:async';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:sellwild_sdk/sellwild_sdk.dart';
import 'package:sellwild_sdk/src/failures/sellwild_log.dart';

import '../flutter_test_config.dart' show blockedSharedClient;
import '../support/contract_schemas.dart';
import '../support/http_mocks.dart';

const int now = 1790000000000;
const String uid = '8F2C1C1E-1B7B-4E0E-9A57-6C3E7C3F4E11';

/// Records what the shell hands to its sink.
class SinkRecorder {
  final List<ClientFailureEvent> events = [];
  final List<bool> flushes = [];

  Future<void> call(ClientFailureEvent event, bool flushNow) async {
    events.add(event);
    flushes.add(flushNow);
  }

  Map<String, String> get attributes => events.single.attributes;
}

class NoMessageException implements Exception {}

class ThrowingToString implements Exception {
  @override
  String toString() => throw StateError('toString failed');
}

SinkRecorder useRecorder({String? partnerCode = 'weatherbug'}) {
  final sink = SinkRecorder();
  SellwildFailures.setContext(
    partnerCode: partnerCode,
    clock: () => now,
    uid: () => uid,
    sink: sink.call,
  );
  return sink;
}

void logListingsHttp() => SellwildFailures.log(
      code: SellwildFailureCode.listingsFetchHttp,
      component: SellwildFailureComponent.listings,
      message: 'HTTP 503',
      httpStatus: 503,
      url: 'https://cache.sellwild.com/listings-img-data-sm?v=2',
      zoneId: '43',
    );

void main() {
  late List<String> printed;

  setUp(() {
    printed = [];
    SellwildLog.printer = printed.add;
  });

  group('SellwildFailures.log', () {
    test('sends one event with the context, uid and clock', () {
      final sink = useRecorder();

      logListingsHttp();

      final event = sink.events.single;
      expect(event.toJson(), {
        'event': 'clientFailure',
        'action': 'listings.fetch.http',
        'label': 'listings',
        'attributes': {
          'code': 'weatherbug',
          'client': 'flutter',
          'clientVersion': sellwildSdkVersion,
          'severity': 'error',
          'fv': '1',
          'msg': 'HTTP 503',
          'httpStatus': '503',
          'host': 'cache.sellwild.com',
          'zoneId': '43',
          'seq': '1',
          'repeat': '1',
        },
        'uid': uid,
        'createdTime': now,
      });
      expect(event.toJson(), conformsTo('client-failure-event'));
      expect(sink.flushes, [true]);
      expect(SellwildFailures.gateState.sessionCount, 1);
    });

    test('the first event of the session and fatal ones flush at once', () {
      final sink = useRecorder();

      logListingsHttp();
      SellwildFailures.log(
          code: SellwildFailureCode.configFetchHttp,
          component: SellwildFailureComponent.remoteConfig);
      SellwildFailures.log(
          code: SellwildFailureCode.configApplyException,
          component: SellwildFailureComponent.configure,
          severity: SellwildFailureSeverity.fatal);

      expect(sink.flushes, [true, false, true]);
    });

    test('the same failure inside a minute is folded into the next one', () {
      final sink = useRecorder();

      logListingsHttp();
      logListingsHttp();

      expect(sink.events, hasLength(1));
      expect(SellwildFailures.gateState.keys.single.suppressed, 1);
    });

    test('wrapper and clientVersion from the context are sent', () {
      final sink = useRecorder();
      SellwildFailures.setContext(wrapper: 'flutter', clientVersion: '9.9.9');

      logListingsHttp();

      expect(sink.attributes['wrapper'], 'flutter');
      expect(sink.attributes['clientVersion'], '9.9.9');
    });

    test('no partner yet sends code unknown', () {
      final sink = useRecorder(partnerCode: null);

      logListingsHttp();

      expect(sink.attributes['code'], 'unknown');
    });
  });

  group('error mapping (FAILURES.md 3.3)', () {
    Map<String, String> logError(Object error, {String? message}) {
      final sink = useRecorder();
      SellwildFailures.log(
        code: SellwildFailureCode.listingsFetchNetwork,
        component: SellwildFailureComponent.listings,
        error: error,
        message: message,
      );
      return sink.attributes;
    }

    test('an exception gives its type name and message', () {
      final cases = <Object, (String, String)>{
        const FormatException('Unexpected character'): (
          'FormatException',
          'Unexpected character'
        ),
        http.ClientException('Connection refused'): (
          'ClientException',
          'Connection refused'
        ),
        TimeoutException('config GET', const Duration(seconds: 5)): (
          'TimeoutException',
          'config GET'
        ),
        const SocketException('Failed host lookup'): (
          'SocketException',
          'Failed host lookup'
        ),
        const HttpException('Connection closed'): (
          'HttpException',
          'Connection closed'
        ),
        SellwildException('HTTP 503 from cache'): (
          'SellwildException',
          'HTTP 503 from cache'
        ),
        StateError('closed'): ('StateError', 'closed'),
      };
      cases.forEach((error, expected) {
        SellwildFailures.resetForTests();
        final attributes = logError(error);
        expect(attributes['errName'], expected.$1, reason: '$error');
        expect(attributes['msg'], expected.$2, reason: '$error');
      });
    });

    test('an exception without a message falls back to toString()', () {
      final noMessage = logError(TimeoutException(null));
      expect(noMessage['errName'], 'TimeoutException');
      expect(noMessage['msg'], 'TimeoutException');

      SellwildFailures.resetForTests();
      final custom = logError(NoMessageException());
      expect(custom['errName'], 'NoMessageException');
      expect(custom['msg'], "Instance of 'NoMessageException'");

      SellwildFailures.resetForTests();
      final empty = logError(const FormatException(''));
      expect(empty['msg'], 'FormatException');
    });

    test('a thrown string is the message, with no type name', () {
      final attributes = logError('socket closed');

      expect(attributes.containsKey('errName'), isFalse);
      expect(attributes['msg'], 'socket closed');
    });

    test('message and error message are joined and sanitized', () {
      final attributes = logError(
        http.ClientException('Failed host lookup: https://jane@x.io/p?q=1'),
        message: 'fetch 1234567 failed',
      );

      expect(attributes['msg'], 'fetch <n> failed: Failed host lookup: x.io');
    });

    test('a throwing toString is counted and the report still goes out', () {
      final attributes = logError(ThrowingToString());

      expect(attributes['errName'], 'ThrowingToString');
      expect(attributes.containsKey('msg'), isFalse);
      expect(SellwildFailures.internalErrorCount, 1);
    });

    test('a very long message is cut before sanitizing; 200 are sent', () {
      final attributes = logError('boom', message: 'x' * 5000);

      expect(attributes['msg']!.runes, hasLength(200));
      expect(attributes['msg'], endsWith('…'));
    });

    test('the core sees at most 1000 units of message and of error', () {
      // 998 digits sanitize to "<n>", so what follows them would be sent.
      // Only the 2 letters within the first 1000 units may reach the core.
      final long = '${'1' * 998}ABCDEFG';

      final attributes = logError(StateError(long), message: long);

      expect(attributes['msg'], '<n>AB');
    });
  });

  group('never throws, never recurses (FAILURES.md 3.4)', () {
    test('a sink that throws is counted', () {
      SellwildFailures.setContext(
          clock: () => now,
          uid: () => uid,
          sink: (event, flushNow) => throw StateError('sink down'));

      expect(logListingsHttp, returnsNormally);
      expect(SellwildFailures.internalErrorCount, 1);
    });

    test('a sink whose future fails is counted once it fails', () async {
      final failed = Completer<void>();
      SellwildFailures.setContext(
          clock: () => now,
          uid: () => uid,
          sink: (event, flushNow) => failed.future);

      logListingsHttp();
      expect(SellwildFailures.internalErrorCount, 0);
      failed.completeError(http.ClientException('offline'));
      await pumpEventQueue();

      expect(SellwildFailures.internalErrorCount, 1);
    });

    test('a clock or uid that throws is counted and nothing is sent', () {
      final sink = useRecorder();
      SellwildFailures.setContext(clock: () => throw StateError('no clock'));

      expect(logListingsHttp, returnsNormally);
      SellwildFailures.setContext(
          clock: () => now, uid: () => throw StateError('no uid'));
      expect(logListingsHttp, returnsNormally);

      expect(sink.events, isEmpty);
      expect(SellwildFailures.internalErrorCount, 2);
      expect(SellwildFailures.gateState.sessionCount, 0);
    });

    test('a nested call from inside log is ignored', () {
      var calls = 0;
      SellwildFailures.setContext(
        clock: () => now,
        uid: () => uid,
        sink: (event, flushNow) async {
          calls++;
          SellwildFailures.log(
              code: SellwildFailureCode.configFetchNetwork,
              component: SellwildFailureComponent.remoteConfig);
        },
      );

      logListingsHttp();

      expect(calls, 1);
      expect(SellwildFailures.gateState.sessionCount, 1);
      // The guard is released afterwards: the next call is decided.
      SellwildFailures.log(
          code: SellwildFailureCode.configFetchNetwork,
          component: SellwildFailureComponent.remoteConfig);
      expect(calls, 2);
    });

    test('a printer that throws during the echo is counted twice at most', () {
      useRecorder();
      SellwildFailures.setContext(debug: true);
      SellwildLog.printer = (line) => throw StateError('printer down');

      expect(logListingsHttp, returnsNormally);

      // Once for the echo, once for the internal-error echo that also threw.
      expect(SellwildFailures.internalErrorCount, 2);
    });
  });

  group('debug echo (FAILURES.md 2)', () {
    test('prints one sanitized line per call only when debug is on', () {
      useRecorder();

      logListingsHttp();
      expect(printed, isEmpty);

      SellwildFailures.setContext(debug: true);
      SellwildFailures.log(
        code: SellwildFailureCode.listingsFetchNetwork,
        component: SellwildFailureComponent.listings,
        message: 'lookup https://cache.sellwild.com/x?uid=jane@example.com',
      );
      logListingsHttp();
      SellwildFailures.log(
          code: 'Not A Code', component: SellwildFailureComponent.listings);

      expect(printed, [
        '[Sellwild] failure listings.fetch.network listings error sent '
            'lookup cache.sellwild.com',
        '[Sellwild] failure listings.fetch.http listings error deduped '
            'HTTP 503',
        '[Sellwild] failure client.code.invalid listings error sent',
      ]);
    });

    test('echoes an internal error by type only', () {
      SellwildFailures.setContext(
          debug: true,
          clock: () => now,
          uid: () => uid,
          sink: (event, flushNow) => throw StateError('sink down'));

      logListingsHttp();

      expect(printed, ['[Sellwild] failure internal-error StateError']);
    });
  });

  group('kill switches (FAILURES.md 5.3, 5.4)', () {
    test('EVENTS_ENABLED off drops even fatal failures', () {
      final sink = useRecorder();
      SellwildFailures.setContext(eventsEnabled: false);

      SellwildFailures.log(
          code: SellwildFailureCode.configFetchHttp,
          component: SellwildFailureComponent.remoteConfig,
          severity: SellwildFailureSeverity.fatal);

      expect(sink.events, isEmpty);
    });

    test('FAILURES_ENABLED "off" drops failures', () {
      final sink = useRecorder();
      SellwildFailures.setContext(failuresEnabled: 'off');

      logListingsHttp();

      expect(sink.events, isEmpty);
    });

    test('sample rate 0 drops all but fatal failures', () {
      final sink = useRecorder();
      SellwildFailures.setContext(failuresSampleRate: 0.0);

      logListingsHttp();
      SellwildFailures.log(
          code: SellwildFailureCode.configFetchHttp,
          component: SellwildFailureComponent.remoteConfig,
          severity: SellwildFailureSeverity.fatal);

      expect(sink.events.single.attributes['severity'], 'fatal');
    });
  });

  group('context', () {
    test('a null argument keeps the current value', () {
      final sink = useRecorder();
      SellwildFailures.setContext(
          eventsEnabled: true, failuresSampleRate: 0.5, debug: true);

      SellwildFailures.setContext(partnerCode: 'antengo');

      final c = SellwildFailures.context;
      expect(c.partnerCode, 'antengo');
      expect(c.eventsEnabled, isTrue);
      expect(c.failuresSampleRate, 0.5);
      expect(c.debug, isTrue);
      expect(c.sink, isNotNull);
      expect(c.clientVersion, sellwildSdkVersion);
      expect(SellwildLog.enabled, isTrue);
      expect(sink.events, isEmpty);
    });

    test('resetForTests clears context, state, counters and SellwildLog', () {
      useRecorder();
      SellwildFailures.setContext(debug: true, sink: (e, f) => throw 'x');
      logListingsHttp();
      expect(SellwildFailures.internalErrorCount, 1);

      SellwildFailures.resetForTests();

      final c = SellwildFailures.context;
      expect(c.partnerCode, isNull);
      expect(c.debug, isFalse);
      expect(c.sink, isNull);
      expect(c.clock, isNull);
      expect(c.uid, isNull);
      expect(SellwildFailures.internalErrorCount, 0);
      expect(SellwildFailures.gateState.sessionCount, 0);
      expect(SellwildLog.enabled, isFalse);
      expect(SellwildLog.printer, isNot(same(printed.add)));
    });
  });

  group('default route: SellwildAPIClient.instance (FAILURES.md 8)', () {
    late HttpRecorder events;

    setUp(() {
      events = HttpRecorder.status(200);
      SellwildAPIClient.instance = SellwildAPIClient(
          client: events.client, uid: 'queue-uid', clock: () => 1);
    });

    tearDown(() => SellwildAPIClient.instance = blockedSharedClient);

    test('posts the event with the queue uid and the wall clock', () async {
      SellwildFailures.setContext(partnerCode: 'weatherbug');
      final before = DateTime.now().millisecondsSinceEpoch;

      logListingsHttp();
      await pumpEventQueue();

      final after = DateTime.now().millisecondsSinceEpoch;
      final request = events.single;
      expect(request.method, 'POST');
      expect(
          request.url.toString(), 'https://events.sellwild.com/events/queue');
      final batch = events.jsonBody() as List;
      expect(batch, conformsTo('events-batch'));
      final event = batch.single as Map<String, dynamic>;
      expect(event['event'], 'clientFailure');
      expect(event['uid'], 'queue-uid');
      expect(event['createdTime'], inInclusiveRange(before, after));
      expect(event['attributes'], {
        'code': 'weatherbug',
        'client': 'flutter',
        'clientVersion': sellwildSdkVersion,
        'severity': 'error',
        'fv': '1',
        'msg': 'HTTP 503',
        'httpStatus': '503',
        'host': 'cache.sellwild.com',
        'zoneId': '43',
        'seq': '1',
        'repeat': '1',
        'type': 'flutter',
        'sdkVersion': sellwildSdkVersion,
      });
    });

    test('the events uid decides sampling', () async {
      // fnv1a32('queue-uid:failures') / 2^32 is about 0.0135: out at a rate
      // of 0.01, in at 0.02.
      SellwildFailures.setContext(failuresSampleRate: 0.01);
      logListingsHttp();
      await pumpEventQueue();
      expect(events.requests, isEmpty);

      SellwildFailures.setContext(failuresSampleRate: 0.02);
      logListingsHttp();
      await pumpEventQueue();
      expect(events.requests, hasLength(1));
    });
  });
}

File: android/src/test/kotlin/com/sellwild/sdk/failures/SellwildFailuresTest.kt
package com.sellwild.sdk.failures

import android.content.Context
import com.sellwild.sdk.SellwildSDK
import com.sellwild.sdk.support.ContractSchemas
import io.mockk.every
import io.mockk.mockk
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import java.io.IOException
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

/** The logFailure shell (FAILURES.md 3.4): context, hold-until-attached, echo, reentrancy, never throws. */
class SellwildFailuresTest {

    @get:Rule
    val failures = FailuresRule()

    private var now = 1_790_000_000_000L
    private val sink = FakeFailureSink()

    private fun attached(): FakeFailureSink {
        SellwildFailures.clock = { now }
        SellwildFailures.bind(sink)
        return sink
    }

    @Test
    fun `log sends one event with the android context and the sink uid`() {
        attached()
        SellwildFailures.setContext { it.copy(partnerCode = "weatherbug") }

        SellwildFailures.log(
            code = SellwildFailureCode.LISTINGS_FETCH_HTTP,
            component = SellwildFailureComponent.LISTINGS,
            message = "HTTP 503",
            httpStatus = 503,
            url = "https://cache.sellwild.com/listings-img-data-sm?v=2",
            zoneId = "43",
        )

        val event = sink.pushed.single()
        assertEquals("clientFailure", event.event)
        assertEquals("listings.fetch.http", event.action)
        assertEquals("listings", event.label)
        assertEquals(VECTOR_UID, event.uid)
        assertEquals(now, event.createdTime)
        assertEquals(
            linkedMapOf(
                "code" to "weatherbug",
                "client" to "android",
                "clientVersion" to SellwildSDK.SDK_VERSION,
                "severity" to "error",
                "fv" to "1",
                "msg" to "HTTP 503",
                "httpStatus" to "503",
                "host" to "cache.sellwild.com",
                "zoneId" to "43",
                "seq" to "1",
                "repeat" to "1",
            ),
            event.attributes,
        )
        assertEquals(listOf(true), sink.flushes)
        assertEquals(1, SellwildFailures.gateState.sessionCount)
        ContractSchemas.assertValid("client-failure-event", event.toJson())
    }

    @Test
    fun `the gate keeps its state between calls`() {
        attached()

        repeat(3) { SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig", message = "HTTP 403") }
        now += FailuresCore.DEDUPE_WINDOW_MS
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig", message = "HTTP 403")
        SellwildFailures.log(code = "config.fetch.parse", component = "remoteConfig")

        assertEquals(listOf("1", "2", "3"), sink.pushed.map { it.attributes["seq"] })
        assertEquals(listOf("1", "3", "1"), sink.pushed.map { it.attributes["repeat"] })
        assertEquals(listOf(true, false, false), sink.flushes)
    }

    @Test
    fun `an error gives errName, message and its first frames without a header`() {
        attached()
        val error = IOException("boom at https://cache.sellwild.com/x?q=1").apply {
            stackTrace = arrayOf(
                StackTraceElement("com.sellwild.sdk.SellwildAPIClient", "fetchListings", "SellwildAPI.kt", 122),
                StackTraceElement("com.sellwild.sdk.SellwildFeedView", "load", "SellwildFeedView.kt", -1),
                StackTraceElement("java.lang.Thread", "run", null, -1),
                StackTraceElement("a.B", "c", "B.kt", 1),
                StackTraceElement("a.B", "d", "B.kt", 2),
                StackTraceElement("a.B", "sixth", "B.kt", 6),
            )
        }

        SellwildFailures.log(code = "listings.fetch.network", component = "listings", error = error, message = "fetch failed")

        val attributes = sink.pushed.single().attributes
        assertEquals("IOException", attributes["errName"])
        assertEquals("fetch failed: boom at cache.sellwild.com", attributes["msg"])
        assertEquals(
            listOf(
                "com.sellwild.sdk.SellwildAPIClient.fetchListings(SellwildAPI.kt:122)",
                "com.sellwild.sdk.SellwildFeedView.load(SellwildFeedView.kt)",
                "java.lang.Thread.run(Unknown Source)",
                "a.B.c(B.kt:1)",
                "a.B.d(B.kt:2)",
            ).joinToString("\n"),
            attributes["stack"],
        )
    }

    @Test
    fun `an anonymous error class sends no errName`() {
        attached()

        SellwildFailures.log(code = "config.fetch.network", component = "remoteConfig", error = object : RuntimeException("x") {})

        assertFalse(sink.pushed.single().attributes.containsKey("errName"))
    }

    @Test
    fun `calls before a sink is attached are held, then decided with their own time and context`() {
        SellwildFailures.clock = { now }
        SellwildFailures.setContext { it.copy(partnerCode = "early") }
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig", httpStatus = 403)
        SellwildFailures.setContext { it.copy(partnerCode = "late", failuresEnabled = false) }
        now += 5_000

        assertEquals(1, SellwildFailures.pendingCount)
        assertTrue(sink.pushed.isEmpty())

        SellwildFailures.bind(sink)

        val event = sink.pushed.single()
        assertEquals("early", event.attributes["code"])
        assertEquals(now - 5_000, event.createdTime)
        assertEquals(VECTOR_UID, event.uid)
        assertEquals(listOf(true), sink.flushes)
        assertEquals(0, SellwildFailures.pendingCount)
    }

    @Test
    fun `held calls are dropped when the events switch set since is off`() {
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig", httpStatus = 403)
        SellwildFailures.setContext { it.copy(eventsEnabled = " OFF ") }

        SellwildFailures.bind(sink)

        assertTrue(sink.pushed.isEmpty())
        assertEquals(0, SellwildFailures.pendingCount)
        assertEquals(FailureState(), SellwildFailures.gateState)
        SellwildFailures.setContext { it.copy(eventsEnabled = 1) }
        SellwildFailures.log(code = "config.fetch.parse", component = "remoteConfig")
        assertEquals("dropped, not held for later", listOf("config.fetch.parse"), sink.pushed.map { it.action })
    }

    @Test
    fun `the hold keeps one session of calls and drops the rest`() {
        SellwildFailures.setContext { it.copy(debug = true) }

        repeat(FailuresCore.SESSION_EMITS + 1) { i ->
            SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig", message = "call $i")
        }

        assertEquals(FailuresCore.SESSION_EMITS, SellwildFailures.pendingCount)
        assertEquals(
            "[Sellwild] failure config.fetch.http remoteConfig error held call 0",
            failures.lines.first(),
        )
        assertEquals(
            "[Sellwild] failure config.fetch.http remoteConfig error held_full call 20",
            failures.lines.last(),
        )
        SellwildFailures.bind(sink)
        assertEquals((0 until FailuresCore.SESSION_EMITS).map { "call $it" }, sink.pushed.map { it.attributes["msg"] })
    }

    @Test
    fun `binding the same sink again replays nothing twice, and a null sink holds calls again`() {
        attached()
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig")
        SellwildFailures.bind(sink)
        assertEquals(1, sink.pushed.size)

        SellwildFailures.bind(null)
        SellwildFailures.log(code = "config.fetch.parse", component = "remoteConfig")

        assertEquals(1, sink.pushed.size)
        assertEquals(1, SellwildFailures.pendingCount)
    }

    @Test
    fun `the debug echo prints one line per call, only when debug is on`() {
        attached()
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig", message = "quiet")
        assertTrue(failures.lines.isEmpty())

        SellwildFailures.setContext { it.copy(debug = true) }
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig", message = "quiet")
        SellwildFailures.log(code = "config.fetch.parse", component = "remoteConfig", message = "jane@example.com")

        assertEquals(
            listOf(
                "[Sellwild] failure config.fetch.http remoteConfig error deduped quiet",
                "[Sellwild] failure config.fetch.parse remoteConfig error sent <email>",
            ),
            failures.lines,
        )
    }

    @Test
    fun `the local failures switch wins over the remote value`() {
        attached()
        SellwildFailures.setContext { it.copy(failuresEnabled = "true", failuresEnabledOverride = false) }
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig")
        assertTrue(sink.pushed.isEmpty())

        SellwildFailures.setContext { it.copy(failuresEnabled = "off", failuresEnabledOverride = true) }
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig")
        assertEquals(1, sink.pushed.size)
    }

    @Test
    fun `remote flags reach the gate raw`() {
        attached()
        SellwildFailures.setContext { it.copy(eventsEnabled = 0) }
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig", severity = SellwildFailureSeverity.FATAL)
        SellwildFailures.setContext { it.copy(eventsEnabled = null, failuresSampleRate = "0") }
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig")
        assertTrue(sink.pushed.isEmpty())

        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig", severity = SellwildFailureSeverity.FATAL)
        assertEquals(listOf("fatal"), sink.pushed.map { it.attributes["severity"] })
    }

    @Test
    fun `setWrapper tags every later event`() {
        attached()
        SellwildFailures.setWrapper("react-native")

        SellwildFailures.log(code = "bridge.props.invalid", component = "bridge")

        assertEquals("react-native", SellwildFailures.context.wrapper)
        assertEquals("react-native", sink.pushed.single().attributes["wrapper"])
    }

    @Test
    fun `a nested call from inside log returns at once`() {
        attached()
        sink.onPush = { SellwildFailures.log(code = "config.fetch.parse", component = "remoteConfig") }

        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig")
        sink.onPush = null
        SellwildFailures.log(code = "config.fetch.timeout", component = "remoteConfig")

        assertEquals(listOf("config.fetch.http", "config.fetch.timeout"), sink.pushed.map { it.action })
        assertEquals(0, SellwildFailures.internalErrorCount)
    }

    @Test
    fun `failures inside log are counted and echoed, never thrown`() {
        attached()
        SellwildFailures.setContext { it.copy(debug = true) }
        sink.onPush = { throw IllegalStateException("sink down") }

        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig")
        SellwildFailures.clock = { throw ArithmeticException("clock down") }
        SellwildFailures.log(code = "config.fetch.parse", component = "remoteConfig")

        assertEquals(2, SellwildFailures.internalErrorCount)
        assertEquals(
            listOf(
                "[Sellwild] failure internal java.lang.IllegalStateException",
                "[Sellwild] failure internal java.lang.ArithmeticException",
            ),
            failures.lines,
        )
    }

    @Test
    fun `a printer that throws is counted for the echo and for the internal echo, and nothing escapes`() {
        attached()
        SellwildFailures.setContext { it.copy(debug = true) }
        SellwildLog.printer = { throw IllegalStateException("printer down") }

        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig")

        assertEquals(2, SellwildFailures.internalErrorCount)
        // The event went out before the echo failed.
        assertEquals(listOf("config.fetch.http"), sink.pushed.map { it.action })
    }

    @Test
    fun `setContext and bind never throw, and one failed replay does not stop the rest`() {
        SellwildFailures.setContext { throw IllegalArgumentException("bad update") }
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig")
        SellwildFailures.log(code = "config.fetch.parse", component = "remoteConfig")
        sink.onPush = { if (it.action == "config.fetch.http") throw IllegalStateException("sink down") }

        SellwildFailures.bind(sink)

        assertEquals(2, SellwildFailures.internalErrorCount)
        assertEquals(listOf("config.fetch.parse"), sink.pushed.map { it.action })
        assertNull(SellwildFailures.context.partnerCode)
    }

    @Test
    fun `held calls the gate drops are not sent`() {
        repeat(2) { SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig", message = "same") }

        SellwildFailures.bind(sink)

        assertEquals(1, sink.pushed.size)
        assertEquals(1, SellwildFailures.gateState.keys.single().suppressed)
    }

    @Test
    fun `a sink whose uid fails is not bound, and the held calls wait for the next one`() {
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig")
        val broken = object : FailureSink {
            override val uid: String get() = throw IllegalStateException("prefs unreadable")
            override fun push(event: FailureEvent, flushNow: Boolean) = Unit
        }

        SellwildFailures.bind(broken)

        assertEquals(1, SellwildFailures.internalErrorCount)
        assertEquals(1, SellwildFailures.pendingCount)
        SellwildFailures.log(code = "config.fetch.parse", component = "remoteConfig")
        assertEquals("still unbound: the next call is held too", 2, SellwildFailures.pendingCount)

        SellwildFailures.bind(sink)

        assertEquals(listOf("config.fetch.http", "config.fetch.parse"), sink.pushed.map { it.action })
        assertEquals(0, SellwildFailures.pendingCount)
    }

    @Test
    fun `the sink uid is read once, when it is bound`() {
        var reads = 0
        val counting = object : FailureSink {
            override val uid: String get() = VECTOR_UID.also { reads++ }
            override fun push(event: FailureEvent, flushNow: Boolean) = sink.push(event, flushNow)
        }
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig")

        SellwildFailures.bind(counting)
        SellwildFailures.log(code = "config.fetch.parse", component = "remoteConfig")
        SellwildFailures.log(code = "config.fetch.timeout", component = "remoteConfig")

        assertEquals(1, reads)
        assertEquals(listOf(VECTOR_UID, VECTOR_UID, VECTOR_UID), sink.pushed.map { it.uid })
    }

    @Test
    fun `attach with a Context that fails is counted, not thrown`() {
        val context = mockk<Context>()
        every { context.applicationContext } throws IllegalStateException("no app context")

        SellwildFailures.attach(context)

        assertEquals(1, SellwildFailures.internalErrorCount)
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig")
        assertEquals("still held: nothing was attached", 1, SellwildFailures.pendingCount)
    }

    @Test
    fun `concurrent calls on other threads are all reported`() {
        attached()
        val pool = Executors.newFixedThreadPool(8)
        val start = CountDownLatch(1)
        try {
            repeat(FailuresCore.SESSION_EMITS) { i ->
                pool.execute {
                    start.await()
                    SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig", message = "thread $i")
                }
            }
            start.countDown()
        } finally {
            pool.shutdown()
            assertTrue(pool.awaitTermination(10, TimeUnit.SECONDS))
        }

        assertEquals(FailuresCore.SESSION_EMITS, sink.pushed.size)
        assertEquals((1..FailuresCore.SESSION_EMITS).map { "$it" }.toSet(), sink.pushed.map { it.attributes["seq"] }.toSet())
    }

    @Test
    fun `resetForTests clears state, context, hold and counters`() {
        SellwildFailures.setContext { it.copy(partnerCode = "p", debug = true) }
        SellwildFailures.log(code = "config.fetch.http", component = "remoteConfig")
        SellwildFailures.setContext { throw IllegalStateException() }
        SellwildFailures.clock = { 1L }

        SellwildFailures.resetForTests()

        assertEquals(SellwildFailureContext(), SellwildFailures.context)
        assertEquals(0, SellwildFailures.pendingCount)
        assertEquals(0, SellwildFailures.internalErrorCount)
        assertEquals(FailureState(), SellwildFailures.gateState)
        assertFalse(SellwildLog.enabled)
        assertTrue(SellwildFailures.clock() > 1L)
    }
}
